### ISSUE TITLE
Elasticsearch Output Tests fail with Elasticsearch 7.5.0

### ISSUE THREAD

**Issue by wagner-certat** [certtools/intelmq#1479]
Running the tests with a local 7.5.0 installation (https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html) fails on the indices part:

```
======================================================================
ERROR: Tests whether an input event with a time.observation field (and no time.source field) is indexed according to
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 200, in test_index_detected_from_time_observation
    "rotate_index": "daily"})
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 249, in base_check_expected_index_created
    self.con.indices.put_template(name=parameters['elastic_index'], body=SAMPLE_TEMPLATE)
  File "/usr/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 76, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/lib/python3.6/site-packages/elasticsearch/client/indices.py", line 447, in put_template
    name), params=params, body=body)
  File "/usr/lib/python3.6/site-packages/elasticsearch/transport.py", line 314, in perform_request
    status, headers, data = connection.perform_request(method, url, params, body, headers=headers, ignore=ignore, timeout=timeout)
  File "/usr/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 161, in perform_request
    self._raise_error(response.status, raw_data)
  File "/usr/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 125, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.RequestError: TransportError(400, 'mapper_parsing_exception', 'Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]')
-------------------- >> begin captured logging << --------------------
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: http://localhost:9200 "HEAD /None-2020-02-02 HTTP/1.1" 404 0
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: http://localhost:9200 "PUT /_template/intelmq HTTP/1.1" 400 942
elasticsearch: WARNING: PUT http://localhost:9200/_template/intelmq [status:400 request:0.013s]
elasticsearch: DEBUG: > {"mappings": {"events": {"properties": {"time.observation": {"type": "date"}, "time.source": {"type": "date"}, "classification.type": {"type": "keyword"}, "source.asn": {"type": "integer"}, "feed.name": {"type": "text"}, "source.ip": {"type": "ip"}}}}, "template": "intelmq-*", "index_patterns": ["intelmq-*"]}
elasticsearch: DEBUG: < {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]"}],"type":"mapper_parsing_exception","reason":"Failed to parse mapping [_doc]: Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]","caused_by":{"type":"mapper_parsing_exception","reason":"Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]"}},"status":400}
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Tests whether an input event with a time.source field is indexed according to its time.source date.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 187, in test_index_detected_from_time_source
    "rotate_index": "daily"})
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 249, in base_check_expected_index_created
    self.con.indices.put_template(name=parameters['elastic_index'], body=SAMPLE_TEMPLATE)
  File "/usr/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 76, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/lib/python3.6/site-packages/elasticsearch/client/indices.py", line 447, in put_template
    name), params=params, body=body)
  File "/usr/lib/python3.6/site-packages/elasticsearch/transport.py", line 314, in perform_request
    status, headers, data = connection.perform_request(method, url, params, body, headers=headers, ignore=ignore, timeout=timeout)
  File "/usr/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 161, in perform_request
    self._raise_error(response.status, raw_data)
  File "/usr/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 125, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.RequestError: TransportError(400, 'mapper_parsing_exception', 'Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]')
-------------------- >> begin captured logging << --------------------
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: http://localhost:9200 "HEAD /None-1869-12-02 HTTP/1.1" 404 0
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: http://localhost:9200 "PUT /_template/intelmq HTTP/1.1" 400 942
elasticsearch: WARNING: PUT http://localhost:9200/_template/intelmq [status:400 request:0.013s]
elasticsearch: DEBUG: > {"mappings": {"events": {"properties": {"time.observation": {"type": "date"}, "time.source": {"type": "date"}, "classification.type": {"type": "keyword"}, "source.asn": {"type": "integer"}, "feed.name": {"type": "text"}, "source.ip": {"type": "ip"}}}}, "template": "intelmq-*", "index_patterns": ["intelmq-*"]}
elasticsearch: DEBUG: < {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]"}],"type":"mapper_parsing_exception","reason":"Failed to parse mapping [_doc]: Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]","caused_by":{"type":"mapper_parsing_exception","reason":"Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]"}},"status":400}
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Tests whether get_index returns an expected default date
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 225, in test_index_falls_back_to_default_date
    parameters=parameters)
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 249, in base_check_expected_index_created
    self.con.indices.put_template(name=parameters['elastic_index'], body=SAMPLE_TEMPLATE)
  File "/usr/lib/python3.6/site-packages/elasticsearch/client/utils.py", line 76, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/usr/lib/python3.6/site-packages/elasticsearch/client/indices.py", line 447, in put_template
    name), params=params, body=body)
  File "/usr/lib/python3.6/site-packages/elasticsearch/transport.py", line 314, in perform_request
    status, headers, data = connection.perform_request(method, url, params, body, headers=headers, ignore=ignore, timeout=timeout)
  File "/usr/lib/python3.6/site-packages/elasticsearch/connection/http_urllib3.py", line 161, in perform_request
    self._raise_error(response.status, raw_data)
  File "/usr/lib/python3.6/site-packages/elasticsearch/connection/base.py", line 125, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.RequestError: TransportError(400, 'mapper_parsing_exception', 'Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]')
-------------------- >> begin captured logging << --------------------
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: http://localhost:9200 "HEAD /intelmq-2018-09-09 HTTP/1.1" 404 0
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: http://localhost:9200 "PUT /_template/intelmq HTTP/1.1" 400 942
elasticsearch: WARNING: PUT http://localhost:9200/_template/intelmq [status:400 request:0.011s]
elasticsearch: DEBUG: > {"mappings": {"events": {"properties": {"time.observation": {"type": "date"}, "time.source": {"type": "date"}, "classification.type": {"type": "keyword"}, "source.asn": {"type": "integer"}, "feed.name": {"type": "text"}, "source.ip": {"type": "ip"}}}}, "template": "intelmq-*", "index_patterns": ["intelmq-*"]}
elasticsearch: DEBUG: < {"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]"}],"type":"mapper_parsing_exception","reason":"Failed to parse mapping [_doc]: Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]","caused_by":{"type":"mapper_parsing_exception","reason":"Root mapping definition has unsupported parameters:  [events : {properties={time.observation={type=date}, time.source={type=date}, classification.type={type=keyword}, source.asn={type=integer}, source.ip={type=ip}, feed.name={type=text}}}]"}},"status":400}
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: Tests whether get_index returns an expected default string
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/tests/bots/outputs/elasticsearch/test_output.py", line 236, in test_index_falls_back_to_default_string
    "rotate_index": "daily"})
  File "/home/sebastian/dev/intelmq/intelmq/lib/test.py", line 234, in prepare_bot
    self.bot = self.bot_reference(self.bot_id)
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 158, in __init__
    self.init()
  File "/home/sebastian/dev/intelmq/intelmq/bots/outputs/elasticsearch/output.py", line 99, in init
    "Have you created the template?".format(self.elastic_index))
RuntimeError: No template with the name 'intelmq' exists on the Elasticsearch host, but 'rotate_index' is set. Have you created the template?
-------------------- >> begin captured logging << --------------------
test-bot: INFO: ElasticsearchOutputBot initialized with id test-bot and intelmq 2.2.0.alpha.1 and python 3.6.9 (default, Oct 29 2019, 10:39:36) [GCC] as process 4883.
test-bot: INFO: Bot is starting.
test-bot: INFO: Looking for Index template 'intelmq'.
urllib3.util.retry: DEBUG: Converted retries value: False -> Retry(total=False, connect=None, read=None, redirect=0, status=None)
urllib3.connectionpool: DEBUG: Starting new HTTP connection (1): 127.0.0.1:9200
urllib3.connectionpool: DEBUG: http://127.0.0.1:9200 "HEAD /_template/intelmq HTTP/1.1" 404 0
test-bot: ERROR: Bot initialization failed.
Traceback (most recent call last):
  File "/home/sebastian/dev/intelmq/intelmq/lib/bot.py", line 158, in __init__
    self.init()
  File "/home/sebastian/dev/intelmq/intelmq/bots/outputs/elasticsearch/output.py", line 99, in init
    "Have you created the template?".format(self.elastic_index))
RuntimeError: No template with the name 'intelmq' exists on the Elasticsearch host, but 'rotate_index' is set. Have you created the template?
test-bot: INFO: Bot stopped.
--------------------- >> end captured logging << ---------------------

----------------------------------------------------------------------
Ran 9 tests in 2.276s

FAILED (errors=4)
```

**Comment by wagner-certat** [comment]
ping @stone-z @navtej @cert-lv

**Comment by navtej** [comment]
tagging @stone-z , is running elasticmapper a pre-requirement? That is my guess.

**Comment by wagner-certat** [comment]
But it works on travis (with elasticsearch 5.5.0) too

**Comment by stone-z** [comment]
The first 3 tests seem to suggest that the Elasticsearch mapping rules changed. What was the last successful ES version? Maybe it is the [deprecation of mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/current/removal-of-types.html) or some other change.

The 4th test looks for a template with that specific name. You don't necessarily need to run `elasticmapper`, but a template has to be there. The reason it looks for the particular name is that there isn't a way (as far as I'm aware) to ask Elasticsearch if a document matches _any_ existing template without actually indexing a document, so rather than let it fail at runtime when the first document is sent, it can complain early. 

That said, I'm not sure why it hasn't been an issue sooner -- that test doesn't create the template but then it runs the bot. The base test `base_check_expected_index_created` demonstrates a way to create the template. I am also happy to take a look but it will be a week or so before I have a chance to get my environment set up and get a look at it.

**Comment by wagner-certat** [comment]
Was anyone able to have a detailed look on it or develop (partial) fixes?

Please let me know if I can help in any kind

**Comment by wagner-certat** [comment]
I open a discussion on this on the mailinglist, the current (and only) proposed solution is abandoning the output bot and recommending logstash: https://lists.cert.at/pipermail/intelmq-dev/2020-March/thread.html

Please speak up *soon* if you do not agree!

ping @stone-z @navtej @cert-lv as potential users of this bot

**Comment by stone-z** [comment]
I like the redis plugin so I think _adding_ the queue for Logstash would be a nice idea, but IMO  abandoning the ES bot altogether would be the wrong way to go. It assumes that Logstash is already in use and can be configured to fetch from a known redis host. In my IntelMQ deployments, Elasticsearch is a long-living production cluster and IntelMQ is more ephemeral, there wouldn't be a static host where Logstash could find it. Elasticsearch is so widely used I think it's worth keeping it as a backend.

That said, I'm able to run the bot just fine with ES 7.6.1, the tests just assume old versions. I have started this work but since I can't install IntelMQ locally at the moment the testing is a little cumbersome for me (very excited for Dockerized IntelMQ).
Basically:
- [x] elasticmapper has been updated for ES 7+
- [x] all references to the index type have gone away
- [ ] remaining references to the elastic doctype should be removed
- [ ] Travis needs a newer ES version

I've opened https://github.com/certtools/intelmq/pull/1513 as a starting point for this effort for ES 7+. Travis has an old version of Elasticsearch, but if updated the tests should look better. These changes aren't backwards-compatible as far as I know -- do you want me to introduce a version parameter for the output bot? Which Elasticsearch versions should still be supported?

**Comment by wagner-certat** [comment]
Thank you for stepping up! Please ping me if you need support or help.

>  Travis needs a newer ES version

I found this in the travis docs: https://docs.travis-ci.com/user/database-setup/#installing-specific-versions-of-elasticsearch

> These changes aren't backwards-compatible as far as I know -- do you want me to introduce a version parameter for the output bot? Which Elasticsearch versions should still be supported?

Dropping support for ES < 7 is ok for me from a maintainer perspective - if someone needs it, she/he needs to speak up and contribute anyway (may sound a bit harsh - but unfortunately it is in this case :/)
Assuming that only ES >= 7 is supported, we do not need to differentiate between versions.

**Comment by stone-z** [comment]
Ok, 7+ is fine with me, will focus on getting the tests working.

I tried those instructions in [a previous run](https://travis-ci.org/github/stone-z/intelmq/jobs/668074924) but haven't done much with Travis so I'm not sure why it fails yet

**Comment by stone-z** [comment]
PR has been updated to include a newer Elasticsearch version (7.6.1) on Travis, removes the `elastic_doctype`, and removes the failing test mentioned in https://github.com/certtools/intelmq/issues/1480. This may still work on some older versions (especially where types had been deprecated and the new functionality was already in place) but no guarantees.
